docs/pages/developer-guide.md

```diff
@@ -936,6 +936,160 @@ In the example above, a server task can be accessed through the address:
+
+The scheduler will generate and store TLS artfiacts for both possible formats (`TLS`, `KEYSTORE`) and changing format won't result in new private key.
+
+Generated artifacts are stored as secrets with following naming scheme:
+
+```
+[DCOS_SPACE]/[pod-index-taskname]__[transport-encryption-name]__[artifact-type]
+```
+
+Example of secrets that would get created based on YAML configuration above:
+
+```
+hello-world/hello-0-server__tls__certificate
+hello-world/hello-0-server__tls__private-key
+hello-world/hello-0-server__tls__root-ca-certificate
+hello-world/hello-0-server__tls__keystore
+hello-world/hello-0-server__tls__truststore
+```
+
+### TLS artifacts secrets lifetime
+
+When `transport-encryption` item is removed from the service configuration (i.e. YAML file) it is **not** removed from the secret store. If the same `transport-encryption` configuration is added back to the service configuration existing TLS artifact will be used.
+
+The `UninstallScheduler` responsible for cleaning up service installation will try to remove all TLS artifact secrets previously provisioned by the scheduler.
+
+### Task artifacts access
+
+Each task that requests one or more TLS artfiacts will get artifacts delivered to the task sandbox directory (`$MESOS_SANDBOX`) as a file-based secret. This feature ensures that data aren't stored on disk and are held only in memory on virtual filesystem. Secret appears as a normal file that can be used in application. A task will get only TLS artifacts that are configured in service configuration.
+
+## Private key and certificate details
+
+A scheduler provisions *private key* and `X.509` *certificate* and exposed them to task in various encodings.
+
+### Private key
+
+Private key is generated by using Java [`KeyPairGenerator`](https://docs.oracle.com/javase/7/docs/api/java/security/KeyPairGenerator.html) initialized with `RSA` algorithm. The `RSA` key is generated with `2048` bit size based on [NIST recommnedations](https://www.keylength.com/en/4/).
+
+### X.509 certificate
+
+A `X.509` certificate for the private key is generated by [DC/OS certificate authority](https://docs.mesosphere.com/1.9/networking/tls-ssl/ca-api/) - `DC/OS CA`. A certificate is signed by DC/OS signing certificate. The signing certificate can be either self-signed root CA certificate created during the cluster installation or user provided signing CA certificate.
+
+A certificate has following subject information:
+
+```
+CN=[pod-index-task].[service-name].autoip.dcos.thisdcos.directory
```

Comment:
I think that's a reasonable compromise. From a quick internet survey, OpenSSL enforces this, but other things do not (since it conflicts with the limits of DNS). But, AFAIK we merely need our CNs to be consistent, yes?